File: Algorithms/lab/lab3/Question2.py
class TrieTreeNode:
    def __init__(self):
        self.children = {}
        self.count = 0

class TrieTree:
    def __init__(self):
        self.root = TrieTreeNode()

    def insert(self, word):
        node = self.root
        for letter in word:
            if letter not in node.children:
                node.children[letter] = TrieTreeNode()
            node = node.children[letter]
        node.count += 1
    
def match(trie_tree, text_len, text):
    count = 0

    for start in range(text_len):
        node = trie_tree.root 
        for next in range(start, text_len):
            if text[next] in node.children:
                node = node.children[text[next]]
                count += node.count
            else:
                break
    return count

if __name__ == "__main__":
    trie_tree = TrieTree()
    word_num, text_len = map(int, input().split())
    words = []
    for i in range(1, word_num + 1):
        len, word = input().split() # len is of no use
        words.append(word)
        trie_tree.insert(word)
    text = input().strip()

    count = match(trie_tree, text_len, text)
    print(count)
    

File: Algorithms/lab/lab3/Question2.c
#include <stdio.h>

#define MAX_WORD_NUM    5000001
#define MAX_WORD_LEN    51
#define MAX_STR_LEN     100001
// next -> Trie Tree, number -> the number of node x
int next[MAX_WORD_NUM][26], number[MAX_WORD_NUM], word_num, text_len, node_id;
char text[MAX_STR_LEN];

void insert(int len, char* word) {
    int node = 0, letter;
    for(int i = 0; i < len; i++){
        letter = word[i] - 'a';
        if(!next[node][letter]) {
            next[node][letter] = ++node_id;
        }
        node = next[node][letter];
    }
    number[node]++;
}

void input() {
    int len; 
    char word[MAX_WORD_LEN];
    scanf("%d %d", &word_num, &text_len);
    for(int i = 0; i < word_num; i++) {
        scanf("%d %s", &len, word);
        insert(len, word);
    }
    scanf("%s", text);
    return;
}

int match() {
    int count = 0;
    for(int ptr = 0; ptr < text_len; ptr++) {
        int node = 0, letter;
        for(int i = ptr; i < text_len; i++) {
            letter = text[i] - 'a';
            node = next[node][letter];
            if(node == 0) {        // no such subsequence
                break;
            }
            count += number[node];
        }
    }
    return count;
}

void print() {
    for(int i = 0; i < node_id; i++) {
        for(int j = 0; j < 4; j++) {
            printf("%d, ", next[i][j]);
        }
        printf("\n");
    }
    printf("\n");
    for(int i = 0; i < node_id; i++) {
        printf("%d, ", number[i]);
    }
    printf("\n");
    return;
}

int main() {
    input();
    // print();
    printf("%d\n", match());
    return 0;
}
